oteapi_dlite/strategies/convert.py

```diff
@@ -0,0 +1,168 @@
+    )
+    property_mappings: bool = Field(
+        False,
+        description="Whether to infer instance from property mappings.",
+    )
+
+
+class DLiteConvertOutputConfig(AttrDict):
+    """Configuration for output instance to generic DLite converter."""
+
+    label: str = Field(
+        None,
+        description="Label to use when storing the instance.",
+    )
+    datamodel: Optional[str] = Field(
+        None,
+        description="URI of data model.  Used for documentation.",
+    )
+
+
+class DLiteConvertStrategyConfig(AttrDict):
+    """Configuration for generic DLite converter."""
+
+    function_name: str = Field(
+        None,
+        description="Name of convert function.  It will be pased the input "
+        "instances as arguments and should return a sequence of output "
+        "instances.",
+    )
+    module_name: str = Field(
+        None,
+        description="Name of Python module containing the convertion function.",
+    )
+    package: Optional[str] = Field(
+        None,
+        description="Used when performing a relative import of the converter "
+        "function.  It specifies the package to use as the anchor point from "
+        "which to resolve the relative import to an absolute import.",
+    )
+    pypi_package: Optional[str] = Field(
+        None,
+        description="Package name on PyPI.  This field is currently only "
+        "informative, but might be used in the future for automatic package "
+        "installation.",
+    )
+    inputs: Sequence[DLiteConvertInputConfig] = Field(
+        None,
+        description="Input instances.",
+    )
+    outputs: Sequence[DLiteConvertOutputConfig] = Field(
+        None,
+        description="Output instances.",
+    )
+
+
+class DLiteConvertConfig(FunctionConfig):
+    """DLite convert strategy resource config."""
+
+    configuration: DLiteConvertStrategyConfig = Field(
+        ..., description="DLite convert strategy-specific configuration."
+    )
+
+
+@dataclass
+class DLiteConvertStrategy:
+    """Generic DLite convert strategy for converting zero or more input
+    instances to zero or more output instances.
+
+    **Registers strategies**:
+
+    - `("functionType", "application/vnd.dlite-convert")`
+
+    """
+
+    convert_config: DLiteConvertConfig
+
+    def initialize(
+        self,
+        session: "Optional[Dict[str, Any]]" = None,
+    ) -> DLiteSessionUpdate:
+        """Initialize."""
+        return DLiteSessionUpdate(collection_id=get_collection(session).uuid)
+
+    def get(
+        self, session: "Optional[Dict[str, Any]]" = None
+    ) -> DLiteSessionUpdate:
+        """Execute the strategy.
+
+        This method will be called through the strategy-specific endpoint
+        of the OTE-API Services.
+
+        Parameters:
+            session: A session-specific dictionary context.
+
+        Returns:
+            SessionUpdate instance.
+        """
+        config = self.convert_config.configuration
```

Comment:
Remember to update the config with the relevant fields from the session

Reply:
Yes, good point. That is needed if we want a later filter to have an effect. 

I was thinking about encouraging not specifying labels in the configuration. But fetching the label from the session could be useful. However, a single `label` in the session will not do, since each input and output has an optional label, so we need to be more specific when specifying labels in the session. 

Would it make sense to allow variable substitutions in the configuration of a partial pipeline, like

```yaml
  https://www.ntnu.edu/physmet/data#image_analyser:
    function:
      functionType: application/vnd.dlite-convert
      configuration:
        module_name: temdata.image_analyser
        function_name: image_analyser
        inputs:
          - datamodel: http://onto-ns.com/meta/0.1/TEMImage
            label: ${temimage}
        outputs:
          - datamodel: http://onto-ns.com/meta/0.1/PrecipitateStatistics
            label: ${precipitate_statistics}
    mapping:
      mappingType: mappings
      prefixes:
       ...
      triples:
        ...
```

where `${temimage}` and `${precipitate_statistics}` are substituted from the session. 

On the *pros* side, this would allow templated partial pipelines with improved re-usability. 

On the *conc* side, it will be an extra layer of complexity.

Reply:
I am not completely sold on the idea of labels. Also, if we start creating declarative partial pipelines with variables, they become hard to exchange/share between instances. Why not refer to a resource directly?

Reply:
A agree that a partial pipeline should document a single data source and in general not contain variables. However, there are cases where fetching parameters from the configuration are useful. One case is a partial pipeline documenting a SQL database. The documentation of the database should be fix, while the query may vary each time we execute the pipeline. Another possible use case is to specify the label a parser should use when storing an newly created instance into the collection and correspondingly, the label a generator should use when fetching an instance from the collection. In this case the labels may be variables when documenting the partial pipelines, but must be assigned and internally consistent before executing the full pipeline. 
Note: often we don't need